files/generate_system_metrics

```diff
@@ -0,0 +1,244 @@
+  # @attr [boolean] verbose Verbose output
+  # @attr [string] hostname Name of the host the metrics are from. In directory name and json file.
+  #
+  class GenerateSystemMetrics
+    #
+    # Initialize class
+    #
+    # @author Randell Pelak
+    #
+    # @param [integer] polling_interval Time in seconds between calls to poll the system for data.
+    # @param [integer] file_interval Time in seconds between the creation of each output file.
+    # @param [string] metric_type cpu|memory
+    # @param [string] metrics_dir The puppet_metrics_collector output directory.
+    # @param [boolean] verbose Verbose output
+    #
+    # @return [void]
+    def initialize(polling_interval, file_interval, metric_type, metrics_dir, verbose = false)
+      @polling_interval = polling_interval
+      @file_interval = file_interval
+      @metric_type = metric_type
+      @metrics_dir = metrics_dir
+      @verbose = verbose
+
+      @hostname = %x[hostname].strip
+      puts "Hostname is: #{@hostname}" if @verbose
+      FileUtils.mkdir_p(@metrics_dir) unless File.directory?(@metrics_dir)
+    end
+
+    # Run sar to collect the raw system data
+    #
+    # @author Randell Pelak
+    #
+    # @return [string] raw output from sar
+    def run_sar
+      times_to_poll = (@file_interval / @polling_interval).round
+      # sar inputs are polling interval and how many times to poll
+      comm_flags = " -r" if @metric_type =~ /memory/
+      comm = "sar #{comm_flags} #{@polling_interval} #{times_to_poll}"
+      puts "sar command is: #{comm}" if @verbose
+      %x[#{comm}]
+    end
+
+    # Parse the sar output and extract the metrics data
+    #
+    # @author Randell Pelak
+    #
+    # @param [array] sar_output
+    #
+    # @raise [RuntimeError] if sar_output doesn't parse correctly
+    #
+    # @return [hash] The metrics data
+    def parse_sar_output(sar_output)
+      sar_output_arr = sar_output.split(/\n+|\r+/).reject(&:empty?).map { |line| line.split }
+
+      headers_line = sar_output_arr.find { |e| e.include? "%user" }
+      sar_error_missing_headers = "sar output invalid or missing headers." +
+                                  "\nFull output:\n#{sar_output}"
+      raise(sar_error_missing_headers) if headers_line.nil?
+      headers = headers_line[2..-1]
+
+      averages_line = sar_output_arr.find { |e| e.include? "Average:" }
+      sar_error_missing_averages = "sar output missing \"Average:\"\nFull output:\n#{sar_output}"
+      raise(sar_error_missing_averages) if averages_line.nil?
+      averages = averages_line[1..-1]
+
+      puts "sar headers and averages:\n#{headers.join(",")}\n#{averages.join(",")}" if @verbose
+
+      sar_error_header_mismatch = "sar output headers and averages columns are not equal." +
+                                  "\nFull output:\n#{sar_output}"
+      raise(sar_error_header_mismatch) unless headers.count == averages.count
+      #combine the arrays into a hash
+      Hash[headers.zip(averages)]
```

Comment:
There are magic numbers for array segments and much validation. The later is fine.  The former is undesirable.

There appear to be two issues that you are addressing.
1. The headers and the averages start with some BS you don't want.
2. Due to the BS, there is a risk of the fields being mismatched.

Given that all of the BS is at the beginning of the headers and the averages field lists, I think that zipping these fields in reverse and tossing away any results that do not have a numeric value solves both problems.

If no data arrays are found for the desired lines (e.g. headers, averages), then the call to `zip` will throw an exception without checking them beforehand. 

With those things in mind, I am wondering if the following refactor may simplify things.  Please feel free to disregard if it does not or is not conservative enough.
```
headers_line = ... # as above
averages_line = ... # as above

# This will zip the data in reverse order to ensure a correct mapping.
# Since it is being converted to a hash, we don't care about order afterwards.
# If either the headers_line or averages_line is nil, then an error will be raised
sar_data = Hash[headers_line.reverse.zip(averages_line.reverse)]

# Now we throw away any data whose value is not numeric.
sar_data = sar_data.select{ |k,v| v  =~ /\A[-+]?[0-9]*\.?[0-9]+\Z/ }

return sar_data
```

Reply:
great idea, I really like it.  implemented.